# Анализ базы данных книжного сервиса
### В данной работе проанализированна база данных сервиса для чтения книг по подписке с целью последующего формулирония ценностного предложения для нового продукта. <a class="tocSkip">
____
##### Предоставлен доступ к базе данных с информацией о книгах, издательствах, авторах, а также пользовательскими обзорами книг. <a class="tocSkip">
____
##### Поставленные задачи: <a class="tocSkip">
1. посчитать, сколько книг вышло после 1 января 2000 года;
2. для каждой книги посчитать количество обзоров и среднюю оценку;
3. определить издательство, которое выпустило наибольшее число книг толще 50 страниц;
4. определите автора с самой высокой средней оценкой книг с 50 и более оценками;
5. посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1">Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.1">Вывод</a></span></li></ul></li><li><span><a href="#Выполнение-поставленных-задач" data-toc-modified-id="Выполнение-поставленных-задач-2">Выполнение поставленных задач</a></span><ul class="toc-item"><li><span><a href="#Посчитать,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитать,-сколько-книг-вышло-после-1-января-2000-года-2.1">Посчитать, сколько книг вышло после 1 января 2000 года</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1.1">Вывод</a></span></li></ul></li><li><span><a href="#Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку-2.2">Для каждой книги посчитать количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-2.3">Определить издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.3.1">Вывод</a></span></li></ul></li><li><span><a href="#Определить-автора-с-самой-высокой-средней-оценкой-книг-с-50-и-более-оценками" data-toc-modified-id="Определить-автора-с-самой-высокой-средней-оценкой-книг-с-50-и-более-оценками-2.4">Определить автора с самой высокой средней оценкой книг с 50 и более оценками</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.4.1">Вывод</a></span></li></ul></li><li><span><a href="#Посчитайть-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитайть-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-2.5">Посчитайть среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.5.1">Вывод</a></span></li></ul></li></ul></li></ul></div>

## Подготовка данных

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': '~~~~',
            'pwd': '~~~~',
            'host': '~~~~',
            'port': ~~~~,
            'db': '~~~~'}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                db_config['pwd'],
                                                db_config['host'],
                                                db_config['port'],
                                                db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
query = ''' SELECT * FROM books '''

books = pd.io.sql.read_sql(query, con = engine)

books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
query = ''' SELECT * FROM authors '''

authors = pd.io.sql.read_sql(query, con = engine)

authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
query = ''' SELECT * FROM publishers '''

publishers = pd.io.sql.read_sql(query, con = engine)

publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [6]:
query = ''' SELECT * FROM ratings '''

ratings = pd.io.sql.read_sql(query, con = engine)

ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
query = ''' SELECT * FROM reviews '''

reviews = pd.io.sql.read_sql(query, con = engine)

reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

### Вывод
- С таблицами и названиями столбцов в них ознакомился.

## Выполнение поставленных задач

### Посчитать, сколько книг вышло после 1 января 2000 года

In [8]:
query1 = ''' 
SELECT
    COUNT(*) AS cnt,
    COUNT(DISTINCT book_id) AS unq_book_ids,
    COUNT(DISTINCT title) AS unq_titles
FROM
    books
WHERE
    publication_date > '2000-01-01';
'''

result1 = pd.io.sql.read_sql(query1, con = engine)

result1

cnt  unq_book_ids  unq_titles
0  819           819         818

#### Вывод
- После 1 января 2000 года вышло 818 книг.
- В таблице есть книга переизданная.

### Для каждой книги посчитать количество обзоров и среднюю оценку

In [9]:
query2 = ''' 
SELECT
    DISTINCT books.title AS title,
    SUM(SUBQ1.reviews_cnt) AS reviews_cnt,
    AVG(SUBQ2.avg_rating) AS avg_rating
FROM
    books
    LEFT JOIN (
        SELECT
            book_id AS book_id,
            COUNT(review_id) AS reviews_cnt
        FROM
            reviews
        GROUP BY
            book_id) AS SUBQ1 ON books.book_id = SUBQ1.book_id
    LEFT JOIN (
        SELECT
            book_id AS book_id,
            AVG(rating) AS avg_rating
        FROM
            ratings
        GROUP BY
            book_id) AS SUBQ2 ON books.book_id = SUBQ2.book_id
GROUP BY
    title;
'''

result2 = pd.io.sql.read_sql(query2, con = engine)

result2.sample(15)

title  reviews_cnt  \
643                                      The Histories          2.0   
677  The Black Cauldron (The Chronicles of Prydain ...          2.0   
0                 Spook: Science Tackles the Afterlife          2.0   
11                                            Saturday          2.0   
862             Winter's Heart (The Wheel of Time  #9)          2.0   
627                        Love in the Time of Cholera          5.0   
445                            Where Are the Children?          2.0   
914  The Constant Princess (The Plantagenet and Tud...          3.0   
323                  Eight Cousins (Eight Cousins  #1)          2.0   
178                                        Salem Falls          3.0   
151  The Hot Zone: The Terrifying True Story of the...          3.0   
19   Rainbow Six (John Clark  #2; Jack Ryan Univers...          2.0   
115                              The Iliad/The Odyssey          1.0   
42   The Closers (Harry Bosch  #11; Harry Bosch Uni...          2.0   
456            By the River Piedra I Sat Down and Wept          3.0   

     avg_rating  
643    3.500000  
677    4.500000  
0      3.000000  
11     3.500000  
862    4.000000  
627    4.083333  
445    4.000000  
914    3.800000  
323    4.500000  
178    4.333333  
151    4.000000  
19     3.500000  
115    4.000000  
42     4.500000  
456    3.333333

### Определить издательство, которое выпустило наибольшее число книг толще 50 страниц

In [10]:
query3 = ''' 
SELECT
    publishers.publisher AS publisher,
    SUBQ1.books_cnt AS books_cnt
FROM
    publishers
    RIGHT JOIN (
        SELECT
            publisher_id AS publisher_id,
            COUNT(book_id) AS books_cnt
        FROM
            books
        WHERE
            num_pages > 50
        GROUP BY
            publisher_id
        ORDER BY
            books_cnt DESC) AS SUBQ1 ON publishers.publisher_id = SUBQ1.publisher_id
ORDER BY
    books_cnt DESC;;
'''

result3 = pd.io.sql.read_sql(query3, con = engine)

result3.head(10)

publisher  books_cnt
0              Penguin Books         42
1                    Vintage         31
2   Grand Central Publishing         25
3           Penguin Classics         24
4                     Bantam         19
5           Ballantine Books         19
6                    Berkley         17
7              Berkley Books         14
8         St. Martin's Press         14
9  William Morrow Paperbacks         13

#### Вывод
- Самое активное издательство - `Penguin Books`.
- Разрыв в количестве изданных книг очень большой.

### Определить автора с самой высокой средней оценкой книг с 50 и более оценками

In [11]:
query4 = ''' 
SELECT
    authors.author AS author,
    AVG(SUBQ2.avg_rating) AS avg_rating
FROM
    authors
    RIGHT JOIN (
        SELECT
            SUBQ1.title AS title,
            SUBQ1.author_id AS author_id,
            AVG(ratings.rating) AS avg_rating
        FROM
            ratings
            LEFT JOIN (
                SELECT
                    author_id AS author_id,
                    title AS title,
                    book_id AS book_id
                FROM
                    books) AS SUBQ1 ON ratings.book_id = SUBQ1.book_id
        GROUP BY
            SUBQ1.title,
            SUBQ1.author_id
        HAVING
            COUNT(rating_id) > 50) AS SUBQ2 ON authors.author_id = SUBQ2.author_id
GROUP BY
    authors.author
ORDER BY
    avg_rating DESC;
'''

result4 = pd.io.sql.read_sql(query4, con = engine)

result4

author  avg_rating
0                          J.K. Rowling/Mary GrandPré    4.283844
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                      J.R.R. Tolkien    4.258446
3                                   Louisa May Alcott    4.192308
4                                       Arthur Golden    4.107143
5                                        Rick Riordan    4.080645
6                                     William Golding    3.901408
7                                       J.D. Salinger    3.825581
8            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
9   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
10                                          Dan Brown    3.754540
11                                         Lois Lowry    3.750000
12            George Orwell/Boris Grabnar/Peter Škerl    3.729730
13                                    Stephenie Meyer    3.662500
14                                     John Steinbeck    3.622951

#### Вывод
- Совершенно не удивительно увидеть Джоан Роулинг на первом месте.

### Посчитайть среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [12]:
query5 = ''' 
SELECT
    AVG(SUBQ2.reviews_cnt) AS avg_reviews_cnt
FROM
    (
    SELECT
        reviews.username AS username,
        SUBQ1.cnt AS ratings_cnt,
        COUNT(reviews.review_id) AS reviews_cnt
    FROM
        reviews
        RIGHT JOIN (
            SELECT
                username AS username,
                COUNT(rating_id) AS cnt
            FROM
                ratings
            GROUP BY
                username

            HAVING
                COUNT(rating_id) > 50
            ORDER BY
                cnt DESC) AS SUBQ1 on reviews.username = SUBQ1.username
    GROUP BY
        reviews.username,
        SUBQ1.cnt) AS SUBQ2;
'''

result5 = pd.io.sql.read_sql(query5, con = engine)

result5

avg_reviews_cnt
0        24.333333

#### Вывод
- Активный читатель не просто ставит много оценок, он ещё и активно обозревает прочитанное.